pkg/gen/nodejs-templates/kind.ts.mustache

```diff
@@ -54,10 +55,37 @@ import { getVersion } from "../../version";
        * @param opts A bag of options that control this resource's behavior.
        */
       constructor(name: string, args?: inputs.{{Group}}.{{Version}}.{{Kind}}, opts?: pulumi.CustomResourceOptions) {
-          const props: pulumi.Inputs = {};
-          {{#Properties}}
+          function unpackNamespace(metadata: inputs.meta.v1.ListMeta | inputs.meta.v1.ObjectMeta) {
+              if (!("namespace" in metadata) || metadata.namespace === undefined) {
+                  return metadata;
+              }
+
+              let namespace: pulumi.Input<string>;
+              if (core.v1.Namespace.isInstance(metadata.namespace)) {
+                  namespace = metadata.namespace.metadata.namespace;
+              } else {
+                  const m = metadata.namespace;
+                  namespace = metadata.namespace;
+              }
+
+              return { ...metadata, namespace };
+          }
+
+          // In Kubernetes, `.metadata.namespace` must be a `string`. In Pulumi, we allow users to
+          // pass in `string | core.v1.Namespace`. Thus, if the user has provided a
+          // `core.v1.Namespace`, we convert it to a `string` representing that namespace's name.
+          if (args !== undefined && args.metadata !== undefined) {
+              args = { ...args, metadata: pulumi.output(args.metadata).apply(unpackNamespace) };
```

Comment:
I have a feeling that there is a subtle issue here if `args.metadata` is an `Output` with a `Namespace` instance passed as `metadata.namespace`. In that case, we would be resolving an `Output` to another `Output`, which does not properly track dependency information due to a bug in the Pulumi SDK. This PR was intended to fix that bug: https://github.com/pulumi/pulumi/pull/2288

While the namespace change would be nice to have, I think it's safer not to take it for 1.0.

Reply:
If we do want to take the namespace change, I think that we can just change the input type signature and leave this up to the marshalling layer, which will marshal the `Namespace` instance as its ID (which is its name).

Reply:
Ok, let's pull it out then.